proto/regen/ecocredit/v1alpha1/tx.proto

```diff
@@ -0,0 +1,167 @@
+    rpc Send(MsgSendRequest) returns (MsgSendResponse);
+
+    // Retire retires a specified number of credits in the holder's account.
+    rpc Retire(MsgRetireRequest) returns (MsgRetireResponse);
+
+    // SetPrecision allows an issuer to increase the decimal precision of a credit batch. It is an experimental feature
+    // to concretely explore an idea proposed in https://github.com/cosmos/cosmos-sdk/issues/7113. The number of decimal
+    // places allowed for a credit batch is determined by the original number of decimal places used with calling CreatBatch.
+    // SetPrecision allows the number of allowed decimal places to be increased, effectively making the supply more
+    // granular without actually changing any balances. It allows asset issuers to be able to issue an asset without needing
+    // to think about how many subdivisions are needed upfront. While it may not be relevant for credits which likely have
+    // a fairly stable market value, I wanted to experiment a bit and this serves as a proof of concept for a broader
+    // bank redesign where say for instance a coin like the ATOM or XRN could be issued in its own units rather than
+    // micro or nano-units. Instead an operation like SetPrecision would allow trading in micro, nano or pico in the future
+    // based on market demand. Arbitrary, unbounded precision is not desirable because this can lead to spam attacks (like
+    // sending 0.000000000000000000000000000001 coins). This is effectively fixed precision so under the hood it is still
+    // basically an integer, but the fixed precision can be increased so its more adaptable long term than just an integer.
+    rpc SetPrecision(MsgSetPrecisionRequest) returns (MsgSetPrecisionResponse);
+}
+
+// MsgCreateClassRequest is the Msg/CreateClass request type.
+message MsgCreateClassRequest {
+
+    // designer is the address of the account which designed the credit class. The designer has special permissions
+    // to change the list of issuers and perform other administrative operations.
+    string designer = 1;
+
+    // issuers are the addresses of the approved issuers.
+    repeated string issuers = 2;
+
+    // metadata is any arbitrary metadata to attached to the credit class.
+    bytes metadata = 3;
+}
+
+// MsgCreateClassResponse is the Msg/CreateClass response type.
+message MsgCreateClassResponse {
+
+    // class_id is the unique ID of the newly created credit class.
+    string class_id = 1;
+}
+
+// MsgCreateBatchRequest is the Msg/CreateBatch request type.
+message MsgCreateBatchRequest {
+
+    // issuer is the address of the batch issuer.
+    string issuer = 1;
+
+    // class_id is the unique ID of the class.
+    string class_id = 2;
+
+    // issuance are the credits issued in the batch.
+    repeated BatchIssuance issuance = 3;
+
+    // metadata is any arbitrary metadata to attached to the credit batch.
+    bytes metadata = 4;
```

Comment:
There is one very important field in all issuance requests: a certificate. I would prefer to have it in the top scope, rather hidden in metadata. Certificate is basically a string with url or a proof. Probably we should even formalize it - and have additional service which will be able to prove the certificates. 

At minimum, I would add a comment with TODO to consider it.

Reply:
let's open a separate issue to discuss metadata. it's a big topic that I'd rather specify off-chain first and then put on-chain. we would need to discuss with Ron on the customer side